File: frontend/src/utils/tramitacaoTexto.js
/** Formatação de descrições na timeline (B9 — parágrafos legíveis). */

const TAG_HTML_RICO = /<\s*(p|div|br|ul|ol|li|h[1-6]|strong|em|span|blockquote)[\s>/]/i;

export function pareceHtmlRico(texto) {
    return TAG_HTML_RICO.test(String(texto || ''));
}

/**
 * @returns {{ modo: 'vazio'|'html'|'texto', html?: string, texto?: string }}
 */
export function descricaoTramitacaoParaExibicao(descricao) {
    const raw = String(descricao || '').trim();
    if (!raw) {
        return { modo: 'vazio' };
    }
    if (pareceHtmlRico(raw)) {
        return { modo: 'html', html: raw };
    }
    return { modo: 'texto', texto: raw.replace(/\r\n/g, '\n') };
}

File: backend/core/services/tramitacao_texto.py
"""Helpers de exibição de descrição de tramitação (B9)."""

from __future__ import annotations

import re

_TAG_HTML_RICO = re.compile(r"<\s*(p|div|br|ul|ol|li|h[1-6]|strong|em|span|blockquote)[\s>/]", re.I)


def parece_html_rico(texto: str | None) -> bool:
    return bool(_TAG_HTML_RICO.search(str(texto or "")))


def descricao_tramitacao_para_exibicao(descricao: str | None) -> dict[str, str]:
    raw = (descricao or "").strip()
    if not raw:
        return {"modo": "vazio"}
    if parece_html_rico(raw):
        return {"modo": "html", "html": raw}
    return {"modo": "texto", "texto": raw.replace("\r\n", "\n")}
